### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by tarcisiofischer**
Add support for Windows and update VC features. Also rerendered with conda-smithy 1.7.0.

### CODE REVIEW

**Comment by tarcisiofischer** [comment]
cc @nicoddemus

File: ci_support/run_docker_build.sh [hunk 1]
```diff
@@ -41,22 +41,7 @@ conda clean --lock
...
-    set -x
-    export CONDA_PY=34
-    set +x
-    conda build /recipe_root --quiet || exit 1
-    upload_or_check_non_existence /recipe_root conda-forge --channel=main || exit 1
-
-    set -x
-    export CONDA_PY=35
-    set +x
+# Embarking on 1 case(s).
```
Comment (jakirkham) [inline ci_support/run_docker_build.sh]: Something is off with this re-rendering as these shouldn't be removed.

**Comment by tarcisiofischer** [comment]
Shouldn't this PR trigger AppVeyor?

**Comment by jakirkham** [comment]
Try changing your `.condarc` to have this and re-render.

```yaml
channels:
  - conda-forge
  - defaults
```

File: ci_support/run_docker_build.sh (hunk 1 above)
Comment (tarcisiofischer) [inline ci_support/run_docker_build.sh]: Additional information:

```
tarcisio@FischerVBox:~/esss_stuff/pyinstaller-feedstock$ ~/miniconda3/bin/conda info
Current conda install:

             platform : linux-64
        conda version : 4.1.12
    conda-env version : 2.6.0
  conda-build version : 2.0.12
       python version : 3.5.2.final.0
     requests version : 2.11.1
     root environment : /home/tarcisio/miniconda3  (writable)
  default environment : /home/tarcisio/miniconda3
     envs directories : /home/tarcisio/miniconda3/envs
        package cache : /home/tarcisio/miniconda3/pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/linux-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/linux-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /home/tarcisio/.condarc
         offline mode : False
    is foreign system : False
```

**Comment by jakirkham** [comment]
It should, but there seems to not be an `appveyor.yml` file. Did that not get committed for some reason?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -55,13 +54,19 @@ requirements:
     - setuptools
     - waf
     - zlib 1.2.*
+    - vc 9  # [win and py27]
+    - vc 10  # [win and py34]
+    - vc 14  # [win and py35]
```
Comment (jakirkham) [inline recipe/meta.yaml]: I think we need something similar for Python 3.6.

Comment (tarcisiofischer) [inline recipe/meta.yaml]: Done. Thanks.

File: appveyor.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    secure: MP4hZYylDyUWEsrt3u3cod2sbFeRwUziH02mvQOdbjsTO/l1yIxDkP/76rSIjcGC
+
+  matrix:
+    - TARGET_ARCH: x86
+      CONDA_PY: 27
+      CONDA_INSTALL_LOCN: C:\\Miniconda
+
+    - TARGET_ARCH: x64
+      CONDA_PY: 27
+      CONDA_INSTALL_LOCN: C:\\Miniconda-x64
```
Comment (tarcisiofischer) [inline appveyor.yml]: Sure.

Comment (tarcisiofischer) [inline appveyor.yml]: https://github.com/conda-forge/waf-feedstock/pull/4

**Comment by nicoddemus** [comment]
@tarcisiofischer would mind add me (and you if you want) as `maintainer` of this package? We might want to help update it in the future.

**Comment by jakirkham** [comment]
Also please enable CircleCI on your fork and repush.

File: appveyor.yml (hunk 3 above)
Comment (tarcisiofischer) [inline appveyor.yml]: @jakirkham , Now that we'll have the waf package on Windows, what should I do to update the appveyor to add the missing entries on the matrix? Should I want for the waf package to publish and then run ```conda smithy rerender```?

**Comment by jakirkham** [comment]
Restarting now that PR ( https://github.com/conda-forge/waf-feedstock/pull/4 ) has been merged. As AppVeyor will only build this after `waf` is built, we can be sure that dependency will be satisfied (barring some intermittent issue).

**Comment by nicoddemus** [comment]
Hmm just noticed that the build matrix on Windows only contains Python 2.7. Should we rerender this feedstock?

**Comment by jakirkham** [comment]
Yeah that is a bit odd. Maybe it had something to do with `waf` not being found? Agreed that we should probably do a re-render. We are on `conda-smithy` 2.0.1 now FYI.

**Comment by tadeu** [comment]
@tarcisiofischer Please merge https://github.com/tarcisiofischer/pyinstaller-feedstock/pull/1, I've fixed the `meta.yaml` so that `conda smithy` will correctly detect Python 3, and I've rerendered it too. Thanks!

**Comment by jakirkham** [comment]
Sorry, what is wrong with the selectors, @tadeu? `py36` is a permissible selector.

**Comment by tadeu** [comment]
@jakirkham I've got this error in `conda smithy 2.0.1`:

```
Error: Invalid selector in meta.yaml line 33:
    - vc14    # [win and py36]
```

Perhaps because it was duplicated?

**Comment by jakirkham** [comment]
It was certainly not supported in older `conda-build`'s, but it is supported in recent versions.

**Comment by tadeu** [comment]
But the error was on `conda-smithy` when trying to rerender

**Comment by jakirkham** [comment]
Which uses `conda-build`. What version of `conda-build` are you using currently? FWICT one would need `2.1.0+` to support this selector.

**Comment by jakirkham** [comment]
Not a problem. Thanks for checking.

**Comment by tadeu** [comment]
You're right, I've just rerendered using `conda build 2.1.2` and there are no changes. Anyway, there's no harm in the `meta.yaml` changes, it is following https://github.com/conda-forge/staged-recipes/wiki/VC-features

**Comment by jakirkham** [comment]
Nope there is not. It was more out of concern that there was a bug in `conda-smithy` not so much a comment on style.

**Comment by tadeu** [comment]
@jakirkham Now there seems to be a problem... I didn't check it earlier, but `appveyor.yml` still contains only the `CONDA_PY: 27` entries.

**Comment by tadeu** [comment]
@jakirkham Could it be because `waf` is not published yet, so `conda smithy` is not being able to resolve it?

**Comment by jakirkham** [comment]
That is a reasonable theory. It seems to be building `waf` as we speak. So maybe another re-rendering can be tried in a few minutes or so.

**Comment by tadeu** [comment]
@jakirkham One think I noticed is that `mtx = special_case_version_matrix(meta, index)` is returning `{(('python', '3.4'),), (('python', '2.7'),)}` only, and then this is being filtered out in the next line (`mtx = list(filter_cases(mtx, ['python >=2.7,<3|>=3.5', 'numpy >=1.11']))`), so only python 2.7 remains.

**Comment by jakirkham** [comment]
That's because `waf` uploaded the Python 3.4 package not that long ago, but hadn't uploaded Python 3.5 yet. It now appears to have uploaded 3.5 and 3.6. So give it another try and see if it still doesn't come up.

**Comment by tadeu** [comment]
Yes, that was the problem, I've rerendered it again now that all `waf` packages are available, thanks @jakirkham.

**Comment by nicoddemus** [comment]
That's what I'm talking about. 😁

**Comment by tadeu** [comment]
@nicoddemus True 😄

**Comment by jakirkham** [comment]
Awesome! Well let's update this PR with a fresh re-rendering and then we can merge it when it passes.

Should add we definitely need to figure out a way for `conda-smithy` to inform us about these missing dependency issues blocking versions of Python, NumPy, etc.. Some discussion on this point in issue ( https://github.com/conda-forge/conda-smithy/issues/411 ) if you are interested.

**Comment by tadeu** [comment]
@jakirkham I didn't know `conda smithy` would silently remove some platforms if it couldn't solve the specs *at rerender time*, even if there's no `skip: True`. Perhaps this should be an error?

**Comment by jakirkham** [comment]
No worries.

**Comment by jakirkham** [comment]
Appears there is some build issue on Windows with `zlib`. Though that is weird as Python is already built against a working copy of `zlib` and we explicitly include `zlib` in this recipe as well. May require some finagling to fix it. IIRC the `LDFLAGS` were added for this purpose on Unix.

Also it looks like `pyinstaller` expects to be able to import `pywintypes`, which it is unable to. IIRC this is part of `pywin32`.  However, we have not packaged `pywin32` yet. ( https://github.com/conda-forge/staged-recipes/issues/1935 ) I believe there is a copy in `defaults` though we should really get the recipe/package into conda-forge.

**Comment by nicoddemus** [comment]
Added conda-forge/staged-recipes#2276, let's see how that goes.

**Comment by nicoddemus** [comment]
Hmm after building `pywin32`, it seems we also need [pefile](https://pypi.python.org/pypi/pefile) in Windows:

```
===== testing package: pyinstaller-3.2-py35_vc14_2 =====
import: 'PyInstaller'
===== pyinstaller-3.2-py35_vc14_2 OK =====
$ pyinstaller --help
Traceback (most recent call last):
  File "C:\Miniconda3-dev\conda-bld\recipe_1485388337637\_t_env\Scripts\pyinstaller-script.py", line 3, in <module>
    import PyInstaller.__main__
  File "C:\Miniconda3-dev\conda-bld\recipe_1485388337637\_t_env\lib\site-packages\PyInstaller\__main__.py", line 21, in <module>
    import PyInstaller.building.build_main
  File "C:\Miniconda3-dev\conda-bld\recipe_1485388337637\_t_env\lib\site-packages\PyInstaller\building\build_main.py", line 34, in <module>
    from .api import PYZ, EXE, COLLECT, MERGE
  File "C:\Miniconda3-dev\conda-bld\recipe_1485388337637\_t_env\lib\site-packages\PyInstaller\building\api.py", line 38, in <module>
    from PyInstaller.utils.win32 import winmanifest, icon, versioninfo, winresource
  File "C:\Miniconda3-dev\conda-bld\recipe_1485388337637\_t_env\lib\site-packages\PyInstaller\utils\win32\versioninfo.py", line 17, in <module>
    import pefile
ImportError: No module named 'pefile'
```

**Comment by nicoddemus** [comment]
Added conda-forge/staged-recipes#2277 for pefile. 😅

**Comment by nicoddemus** [comment]
After building both `pywin32` and `pefile` locally, finally:

```
TEST END: pyinstaller-3.2-py35_vc14_2
INFO conda_build.config:__exit__(405): --dirty flag not specified.  Removing build folder after successful build/test.

# If you want to upload this package to anaconda.org later, type:
#
# $ anaconda upload C:\Miniconda3-dev\conda-bld\win-64\pyinstaller-3.2-py35_vc14_2.tar.bz2
#
# To have conda build upload to anaconda.org automatically, use
# $ conda config --set anaconda_upload yes
```

🎆

**Comment by nicoddemus** [comment]
That is, after updating the recipe accordingly:

```diff
diff --git a/recipe/meta.yaml b/recipe/meta.yaml
index 9d2d08e..995d629 100644
--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -61,7 +61,9 @@ requirements:
   run:
     - python
     - setuptools
+    - pefile  # [win]
     - pycrypto
+    - pywin32  # [win]
     - zlib 1.2.*
     - vc 9  # [win and py27]
     - vc 10  # [win and py34]
```

**Comment by jakirkham** [comment]
Thanks @nicoddemus. Sorry for sending you down the path to dependency hell.

**Comment by nicoddemus** [comment]
All part of the game I guess. 😄

**Comment by nicoddemus** [comment]
I think we can reopen this PR now that `waf` is built on Windows. 😁

**Comment by nicoddemus** [comment]
Hmm for some reason it failed on `py36`, not being able to import `pywintypes`. 😭 

I can reproduce the problem:

```
$ conda create -n pywin-test-3.6 python=3.6 pywin32
$ activate pywin-test-3.6
$ python -c "import pywintypes"
python -c "import pywintypes"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "c:\Miniconda3-dev\envs\pywin-test-3.6\lib\site-packages\win32\lib\pywintypes.py", line 124, in <module>
    __import_pywin32_system_module__("pywintypes", globals())
  File "c:\Miniconda3-dev\envs\pywin-test-3.6\lib\site-packages\win32\lib\pywintypes.py", line 98, in __import_pywin32_system_module__
    raise ImportError("No system module '%s' (%s)" % (modname, filename))
ImportError: No system module 'pywintypes' (pywintypes36.dll)
```

While everything works fine on 3.5:

```
$ conda create -n pywin-test-3.5 python=3.6 pywin32
$ activate pywin-test-3.5
$ python -c "import pywintypes"
```

Will investigate some more.

**Comment by nicoddemus** [comment]
Hmmm the weirdness continues. If I get the [run_test.py]() file from `pywin32`'s feedstock, it fails on my testing environment for 3.6, and works for 3.5:


```
$ activate pywin-test-3.6
$ python x:\pywin-test-src\run_test.py
Traceback (most recent call last):
  File "x:\pywin-test-src\run_test.py", line 9, in <module>
    import mmapfile
ImportError: DLL load failed: Não foi possível encontrar o módulo especificado.
```

```
$ activate pywin-test-3.5
$ python x:\pywin-test-src\run_test.py
```

Using the `depends` utility I can see that `mmapfile` needs `pywintypes36.dll` to be imported. How did the recipe testing stage of `pywin32` passed, I wonder?

**Comment by nicoddemus** [comment]
Both DLLs are located in the same place: `Lib\site-packages\pywin32_system32`.

**Comment by nicoddemus** [comment]
Oh, on py35 there's a `pywintypes35.dll` next to `python.exe`, that's the dll which is found by `pywintypes.py` on py35. There's now `pywintypes36.dll` next to `python.exe` on my Python 3.6 environment.

**Comment by nicoddemus** [comment]
Oh my that's embarrassing, all those tests were using `pywin32` from Anaconda. 😓 

Let me remake them.

**Comment by nicoddemus** [comment]
Recreated the environments, now both of them fail in the same way:

**3.5**:

```
(pywin-test-3.5) C:\Users\Bruno>conda list
# packages in environment at c:\Miniconda3-dev\envs\pywin-test-3.5:
#
certifi                   2017.1.23                py35_0    conda-forge
pip                       9.0.1                    py35_0    conda-forge
python                    3.5.3                         0    conda-forge
pywin32                   220                      py35_0    conda-forge
setuptools                33.1.0                   py35_0    conda-forge
vs2015_runtime            14.0.25420                    0    conda-forge
wheel                     0.29.0                   py35_0    conda-forge
wincertstore              0.2                      py35_0    conda-forge

(pywin-test-3.5) C:\Users\Bruno>python
Python 3.5.3 | packaged by conda-forge | (default, Jan 23 2017, 20:04:35) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> ^Z


(pywin-test-3.5) C:\Users\Bruno>python -c "import pywintypes"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "c:\Miniconda3-dev\envs\pywin-test-3.5\lib\site-packages\win32\lib\pywintypes.py", line 124, in <module>
    __import_pywin32_system_module__("pywintypes", globals())
  File "c:\Miniconda3-dev\envs\pywin-test-3.5\lib\site-packages\win32\lib\pywintypes.py", line 98, in __import_pywin32_system_module__
    raise ImportError("No system module '%s' (%s)" % (modname, filename))
ImportError: No system module 'pywintypes' (pywintypes35.dll)
```

**3.6**:

```
(pywin-test-3.6) C:\Users\Bruno>conda list
# packages in environment at c:\Miniconda3-dev\envs\pywin-test-3.6:
#
certifi                   2017.1.23                py36_0    conda-forge
pip                       9.0.1                    py36_0    conda-forge
python                    3.6.0                         1    conda-forge
pywin32                   220                      py36_0    conda-forge
setuptools                33.1.0                   py36_0    conda-forge
vs2015_runtime            14.0.25420                    0    conda-forge
wheel                     0.29.0                   py36_0    conda-forge
wincertstore              0.2                      py36_0    conda-forge

(pywin-test-3.6) C:\Users\Bruno>python
Python 3.6.0 | packaged by conda-forge | (default, Jan 14 2017, 00:31:55) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> ^Z


(pywin-test-3.6) C:\Users\Bruno>python -c "import pywintypes"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "c:\Miniconda3-dev\envs\pywin-test-3.6\lib\site-packages\win32\lib\pywintypes.py", line 124, in <module>
    __import_pywin32_system_module__("pywintypes", globals())
  File "c:\Miniconda3-dev\envs\pywin-test-3.6\lib\site-packages\win32\lib\pywintypes.py", line 98, in __import_pywin32_system_module__
    raise ImportError("No system module '%s' (%s)" % (modname, filename))
ImportError: No system module 'pywintypes' (pywintypes36.dll)
```

On might conclude that there's a problem with `pywin32` package, although I can't figure out how the build of [`pywin32` was successful](https://ci.appveyor.com/project/conda-forge/pywin32-feedstock/branch/master).

**Comment by nicoddemus** [comment]
Hmmm both pyinstaller builds seemed to use the same `pywin32` package from conda-forge:

[3.5](https://ci.appveyor.com/project/conda-forge/pyinstaller-feedstock/build/1.0.8/job/kptqibdpqxxdpesg)
```
    pywin32:        220-py35_0       conda-forge
```

[3.6](https://ci.appveyor.com/project/conda-forge/pyinstaller-feedstock/build/1.0.8/job/rhki4038x9q2vm4u#L1838)

```
    pywin32:        220-py36_0       conda-forge
```

☹️ 

If I copy the `pywintypes35.dll` and `pywintypes36.dll` to their Python root directory, both imports now work.

**Comment by nicoddemus** [comment]
@tarcisiofischer the new [pywin32](https://anaconda.org/conda-forge/pywin32/files) build has been published, could you please open/reopen the PR again? Thanks

**Comment by nicoddemus** [comment]
Windows passed! 🎆 😌

**Comment by tarcisiofischer** [comment]
Awesome! :D :fireworks:

**Comment by jakirkham** [comment]
Restarted Travis CI during the weekend lull and it passed as well. Thanks guys. Let's get this in.

**Pull request merged**